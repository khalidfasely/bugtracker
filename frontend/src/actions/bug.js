import bugApi from "../api/bug";

export const setBug = (bug) => ({
    type: "SET_BUG",
    bug
});

export const startSetBug = (pid, bid) => {
    return (dispatch) => {
        return bugApi(pid, bid).then((result) => {
            dispatch(setBug(result.bug));
            return result;
        });
    };
};