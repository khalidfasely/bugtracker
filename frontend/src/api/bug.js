export default async (pid, bid) => {
    const res = await fetch(`/api/project/${pid}/bug/${bid}`);

    const result = await res.json();

    return result;
};