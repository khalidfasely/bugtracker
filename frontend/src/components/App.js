function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
