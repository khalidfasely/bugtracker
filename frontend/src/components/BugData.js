const BugData = ({ bug }) => (
    <div className="bug-data">
        <div className="bug-data__title-desc">
            <div>
                <div className="active-point">
                    <abbr title={bug?.active ? 'Active' : 'Not Active'}>
                        <div className={`is-active ${bug?.active ? 'active' : 'not-active'}`}></div>
                    </abbr>
                </div>
                <abbr title={`Priority: ${bug?.classification}`}>
                    <h3 className={`bug__title ${bug?.classification}`}>{bug?.title}</h3>
                </abbr>
            </div>
            <p className="bug__description">{bug?.description}</p>
        </div>
        <div className="bug-data__user-time">
            <p className="bug__user"><span className="static-text">By: </span>{bug?.user?.username}</p>
            <p className="bug__time"><span className="static-text">On: </span>{bug?.time}</p>
        </div>
        <div>
            <span className="static-text">Users in: </span>{bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <div>
        <span className="static-text">Admins in: </span>{bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;