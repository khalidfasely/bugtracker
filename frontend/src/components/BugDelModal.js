import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;