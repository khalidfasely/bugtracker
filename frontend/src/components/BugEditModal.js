import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;