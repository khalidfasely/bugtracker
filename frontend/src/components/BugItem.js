import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";
import BugData from "./BugData";
import NewComment from "./NewComment";
import { startSetDeleteBug } from "../actions/project";
import { startSetSelectUsers } from '../actions/selectUsers';
import BugDelModal from "./BugDelModal";
import BugEditModal from "./BugEditModal";

export const BugItem = ({
        uname, bug, comments, startSetDeleteBug,
        users, startSetSelectUsers
    }) => {
    const history = useNavigate();

    const [delModalOpen, setDelModalOpen] = useState(false);
    const [editModalOpen, setEditModalOpen] = useState(false);

    // Set select users
    useEffect(() => startSetSelectUsers(bug.on_project), []);

    const deleteBug = () => {
        startSetDeleteBug(bug.id).then(() => {
            history(`/project/${bug.on_project}`);
            setDelModalOpen(false);
        });
    };

    return (
        <div>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    <div className="bug-infos">
                        <BugData bug={bug} />
                        {
                            uname===bug.user.username &&
                            <div className="bug-buttons">
                                <button onClick={() => setEditModalOpen(true)}>Edit</button>
                                <button onClick={() => setDelModalOpen(true)}>Delete</button>
                            </div>
                        }
                    </div>
                    <NewComment bugId={bug.id} />
                    {
                        comments.length > 0 ?
                        comments?.map(comment => (
                            <div key={comment.id}><Comment comment={comment} /></div>
                        )) :
                        <div className="bug-empty"><p>No comments now!</p></div>
                    }
                </div> :
                <div>Not able to see this bug! <Link to={`/project/${bug.on_project}`}>Create one!</Link></div>
            }
            <BugDelModal
                delModalOpen={delModalOpen}
                setDelModalOpen={setDelModalOpen}
                deleteBug={deleteBug}
            />
            <BugEditModal
                editModalOpen={editModalOpen}
                setEditModalOpen={setEditModalOpen}
                users={users}
                bug={bug}
            />
        </div>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments,
    users: state.selectUsers
});

const mapDispatchToProps = (dispatch) => ({
    startSetDeleteBug: (bid) => dispatch(startSetDeleteBug(bid)),
    startSetSelectUsers: (pid) => dispatch(startSetSelectUsers(pid))
});

export default connect(mapStateToProps, mapDispatchToProps)(BugItem);