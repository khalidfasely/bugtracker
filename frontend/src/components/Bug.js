import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);