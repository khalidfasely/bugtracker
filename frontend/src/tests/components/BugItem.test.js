import React from "react";
import { render } from "@testing-library/react";

import { BugItem } from "../../components/BugItem";
import bug from '../fixtures/bug';
import comments from '../fixtures/comments';

jest.mock('../../components/BugDelModal', () => () => <div>BugDelModal</div>);
jest.mock('../../components/BugEditModal', () => () => <div>BugEditModal</div>);
jest.mock('../../components/NewComment', () => () => <div>NewComment</div>);
jest.mock('../../components/Comment', () => () => <div>Comment</div>);

let props;

beforeEach(() => {
    props = {
        uname: 'User',
        bug,
        comments,
        startSetDeleteBug: jest.fn(() => Promise.resolve()),
        startSetSelectUsers: jest.fn()
    }
});

test('Should render BugItem component correctly with uname not in the bug', () => {
    const { asFragment } = render(<BugItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render BugItem component correctly with uname in the bug', () => {
    const { asFragment } = render(<BugItem {...{...props, uname: '1'}} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render BugItem component correctly with uname === bug.user', () => {
    const { asFragment } = render(<BugItem {...{...props, uname: bug.user.username}} />);
    expect(asFragment()).toMatchSnapshot();
});