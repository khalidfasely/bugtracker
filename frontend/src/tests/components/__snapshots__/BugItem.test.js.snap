// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render BugItem component correctly with uname === bug.user 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="bug-infos"
      >
        <div
          class="bug-data"
        >
          <div
            class="bug-data__title-desc"
          >
            <div>
              <div
                class="active-point"
              >
                <abbr
                  title="Active"
                >
                  <div
                    class="is-active active"
                  />
                </abbr>
              </div>
              <abbr
                title="Priority: medium"
              >
                <h3
                  class="bug__title medium"
                >
                  Bug title
                </h3>
              </abbr>
            </div>
            <p
              class="bug__description"
            >
              Bug description
            </p>
          </div>
          <div
            class="bug-data__user-time"
          >
            <p
              class="bug__user"
            >
              <span
                class="static-text"
              >
                By: 
              </span>
              admin
            </p>
            <p
              class="bug__time"
            >
              <span
                class="static-text"
              >
                On: 
              </span>
              now
            </p>
          </div>
          <div>
            <span
              class="static-text"
            >
              Users in: 
            </span>
            <span>
              admin 
            </span>
            <span>
              1 
            </span>
          </div>
          <div>
            <span
              class="static-text"
            >
              Admins in: 
            </span>
            <span>
              admin 
            </span>
            <span>
              1 
            </span>
          </div>
        </div>
        <div
          class="bug-buttons"
        >
          <button>
            Edit
          </button>
          <button>
            Delete
          </button>
        </div>
      </div>
      <div>
        NewComment
      </div>
      <div>
        <div>
          Comment
        </div>
      </div>
      <div>
        <div>
          Comment
        </div>
      </div>
    </div>
    <div>
      BugDelModal
    </div>
    <div>
      BugEditModal
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render BugItem component correctly with uname in the bug 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="bug-infos"
      >
        <div
          class="bug-data"
        >
          <div
            class="bug-data__title-desc"
          >
            <div>
              <div
                class="active-point"
              >
                <abbr
                  title="Active"
                >
                  <div
                    class="is-active active"
                  />
                </abbr>
              </div>
              <abbr
                title="Priority: medium"
              >
                <h3
                  class="bug__title medium"
                >
                  Bug title
                </h3>
              </abbr>
            </div>
            <p
              class="bug__description"
            >
              Bug description
            </p>
          </div>
          <div
            class="bug-data__user-time"
          >
            <p
              class="bug__user"
            >
              <span
                class="static-text"
              >
                By: 
              </span>
              admin
            </p>
            <p
              class="bug__time"
            >
              <span
                class="static-text"
              >
                On: 
              </span>
              now
            </p>
          </div>
          <div>
            <span
              class="static-text"
            >
              Users in: 
            </span>
            <span>
              admin 
            </span>
            <span>
              1 
            </span>
          </div>
          <div>
            <span
              class="static-text"
            >
              Admins in: 
            </span>
            <span>
              admin 
            </span>
            <span>
              1 
            </span>
          </div>
        </div>
      </div>
      <div>
        NewComment
      </div>
      <div>
        <div>
          Comment
        </div>
      </div>
      <div>
        <div>
          Comment
        </div>
      </div>
    </div>
    <div>
      BugDelModal
    </div>
    <div>
      BugEditModal
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render BugItem component correctly with uname not in the bug 1`] = `
<DocumentFragment>
  <div>
    <div>
      Not able to see this bug! 
      <a>
        Create one!
      </a>
    </div>
    <div>
      BugDelModal
    </div>
    <div>
      BugEditModal
    </div>
  </div>
</DocumentFragment>
`;
